Add/lifecycle heartbeat

hyeong01 opened the pull request:
**Issue #, if available:**
#493 NTH should issue lifecycle heartbeats

**Description of changes:**
1. Added the option to issue heartbeats
- Configurable interval and length 
2. Added heartbeat configuration validity check
3. Added automatic closure when lifecycle action becomes invalid
4. Added unit and E2E testing
5. Added explanation of the feature to readme

**How you tested your changes:**
Environment (Linux / Windows): Linux
Kubernetes Version: v1.31.3

Unit testing:
- Checked heartbeat signals and closure due to heartbeat expiration
- Checked heartbeat signals and closure due to drain completion
- Checked heartbeat failures due to invalid lifecycle hook target
- Checked heartbeat failures due to not invalid lifecycle hook target

E2E testing:
- E2E tested heartbeat signal and closure due to heartbeat expiration with kind and localstack. 

Manual testing with real ASG and K8s:
- Tested all 32 possible configuration cases 
- Tested interval > timeout case
- Graceful termination 5->3, 5->4, 10->3, 103->3 (corresponding number of workers for terminating instances) with varying intervals and length for each.
- Important example 1: # of terminations=103->3, Interval=300, heartbeatUntil=590, timeout=590. Last interruption event entering the processInterruptionEvent took 223 seconds since the first event was stored to the eventStore. The first heartbeat was sent out 90 seconds after the last interruption event entering. 
- Important example 2: # of terminations=103->3, Interval=90, heartbeatUntil=260, timeout=150. The timeout was not enough for the last heartbeat to be sent out. 
- Ungraceful termination 5->3, 10->3, 20->3, 103->3 (smaller number of workers than terminating instances).

**Potential Improvements:**
Cached API call for getting heartbeat timeout from ASG (`describeLifecycleHook`).
- Details: 1 hour cache for retrieving the heartbeat timeout value and warning user if timeout < interval.
- Pros: Reduced number of API calls. 
- Cons: Complexity of the system. Cache expires or not used often because termination events happen sparsely.

Single heartbeat manager (single gorountine for issuing heartbeats)
- Pros: Reduced number of goroutines
- Cons: Potential burst in API calls. goroutine is light and the number of goroutines does not grow more than the number of workers (default 10).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Lu-David commented:
Regarding potential improvements: I think caching describeLifecycleHook is unnecessary because like you said termination events are not that frequent and would add unnecessary complexity. Regarding single heartbeat manager, I think it also adds unnecessary complexity

Lu-David commented on pkg/monitor/sqsevent/asg-lifecycle-event.go:
```diff
@@ -95,26 +98,117 @@ func (m SQSMonitor) asgTerminationToInterruptionEvent(event *EventBridgeEvent, m
+
+	for {
+		select {
+		case <-stopCh:
```
What's the purpose of stopCh that isn't already covered by timeout?

Lu-David replied:
oh actually it makes sense

Lu-David commented on README.md:
```diff
@@ -612,6 +612,71 @@ extraScrapeConfigs: |
+
+### Important Notes
+
+- A lifecycle hook for instance termination is required for this feature. Longer grace periods are achieved by renewing the heartbeat timeout of the ASG's lifecycle hook. Instances terminate instantly without a hook.
```
I think this section is important enough that it could be described in the How to use section? The How to use section currently seems sparse.

hyeong01 replied:
Updated this part with detailed and extensive explanation

Lu-David commented on test/e2e/asg-lifecycle-sqs-heartbeat-test:
```diff
@@ -0,0 +1,222 @@
+ASG_TERMINATE_EVENT_ONE_LINE=$(echo "${ASG_TERMINATE_EVENT}" | tr -d '\n' |sed 's/\"/\\"/g')
+SEND_SQS_CMD="awslocal sqs send-message --queue-url ${queue_url} --message-body \"${ASG_TERMINATE_EVENT_ONE_LINE}\" --region ${AWS_REGION}"
+kubectl exec -i "${localstack_pod}" -- bash -c "$SEND_SQS_CMD"
+echo "✅ Sent Spot Interruption Event to SQS queue: ${queue_url}"
```
nit: ASG termination event

Lu-David commented on test/e2e/asg-lifecycle-sqs-heartbeat-test:
```diff
@@ -0,0 +1,222 @@
+    if [[ $FOUND_HEARTBEAT_END_LOG -eq 0 ]] && kubectl logs -n kube-system "${NTH_POD}" | grep -q "Heartbeat deadline exceeded, stopping heartbeat"; then
+        FOUND_HEARTBEAT_END_LOG=1
+    fi
+    if [[ $HEARTBEAT_COUNT -eq 3 && $FOUND_HEARTBEAT_END_LOG -eq 1 ]]; then
```
nit: Instead of hardcoding these values, can we abstract the values of HEARTBEAT_INTERVAL and HEARTBEAT_UNTIL to figure out expected number of Heartbeats at the top of the file? Just to make it easier to update this test file if need be

hyeong01 replied:
What is your opinion about auto calculating the HEARTBEAT_CHECK_CYCLES and HEARTBEAT_CHECK_SLEEP as well? They also derived from HEARTBEAT_INTERVAL and HEARTBEAT_UNTIL, it might be more consistent to either auto-calculate all related values none of them. This would make the code easier to understand?

Lu-David replied:
yeah I think that would be helpful thanks

Lu-David commented on pkg/monitor/sqsevent/asg-lifecycle-event.go:
```diff
@@ -95,26 +98,117 @@ func (m SQSMonitor) asgTerminationToInterruptionEvent(event *EventBridgeEvent, m
+	heartbeatTimeout := int(*lifecyclehook.LifecycleHooks[0].HeartbeatTimeout)
+
+	if heartbeatInterval >= heartbeatTimeout {
+		log.Warn().Msgf("Heartbeat interval (%d seconds) is equal to or greater than the heartbeat timeout (%d seconds) for the lifecycle hook %s. The node would likely be terminated before the heartbeat is sent", heartbeatInterval, heartbeatTimeout, *lifecyclehook.LifecycleHooks[0].LifecycleHookName)
```
can we also add the ASG name (lifecycleDetail.AutoScalingGroupName) in this log warn? Just to help with debugging?

LikithaVemulapalli reviewed (changes requested):
Left some inline comments, good work @hyeong01 !

hyeong01 commented on README.md:
```diff
@@ -612,6 +612,71 @@ extraScrapeConfigs: |
 In IMDS mode, metrics can be collected as follows:
 - Use a `podMonitor` custom resource with the Prometheus Operator to collect metrics.
 
+## Issuing Lifecycle Heartbeats
```
I can add a row in the matrix, but regarding putting the whole section under the SQS mode, I was worried about making it too long. What do you think?

LikithaVemulapalli replied:
If Heartbeats are only supported in SQS, good to keep it within SQS and include it in matrix, even when user reads they can figure out that it only works with SQS without actual trying sending hearbeats

hyeong01 commented on README.md:
```diff
@@ -612,6 +612,71 @@ extraScrapeConfigs: |
+
+### How to use
+
+- Specify values for `Heartbeat Interval` (required) and `Heartbeat Until` (optional).
```
That is an important point. I have written a thorough explanation for this. If written explanation is not sufficient / too long I can change it to a table.

LikithaVemulapalli replied:
Not table but having them as points will be helpful :)

hyeong01 commented on pkg/monitor/sqsevent/sqs-monitor_test.go:
```diff
@@ -908,6 +907,130 @@ func TestMonitor_InstanceNotManaged(t *testing.T) {
+
+	sqsMonitor.SendHeartbeats(1, 6, &lifecycleDetail, stopHeartbeatCh)
+
+	h.Assert(t, h.HeartbeatCallCount == 1, "1 Heartbeat Expected, got %d", h.HeartbeatCallCount)
```
Extracted to helper method & extended the scope of uniting testing to include the whole heartbeat process.

hyeong01 commented on pkg/monitor/sqsevent/asg-lifecycle-event.go:
```diff
@@ -95,26 +98,117 @@ func (m SQSMonitor) asgTerminationToInterruptionEvent(event *EventBridgeEvent, m
 
 	interruptionEvent.PreDrainTask = func(interruptionEvent monitor.InterruptionEvent, n node.Node) error {
+		nthConfig := n.GetNthConfig()
+		if nthConfig.HeartbeatInterval != -1 && nthConfig.HeartbeatUntil != -1 {
```
I missed this comment. Will add a comment right now.

hyeong01 commented:
On top of the changes suggested from the comment, I modified the logging to reduce duplicate logs and added a check that prevents `hearbeatUntil` < `heartbeatInterval`.

LikithaVemulapalli commented on test/e2e/asg-lifecycle-sqs-heartbeat-test:
```diff
@@ -0,0 +1,228 @@
+    fail_and_exit 2
+fi
+
+ASG_TERMINATE_EVENT=$(cat <<EOF
```
nit: ASG_TERMINATION_EVENT

LikithaVemulapalli commented on test/e2e/asg-lifecycle-sqs-heartbeat-test:
```diff
@@ -0,0 +1,228 @@
+EOF
+)
+
+ASG_TERMINATE_EVENT_ONE_LINE=$(echo "${ASG_TERMINATE_EVENT}" | tr -d '\n' |sed 's/\"/\\"/g')
```
same as above `TERMINATION`

LikithaVemulapalli reviewed (changes requested):
left some comments

Lu-David commented on README.md:
```diff
@@ -81,6 +81,78 @@ When using the ASG Lifecycle Hooks, ASG first sends the lifecycle action notific
+#### Issuing Lifecycle Heartbeats
+
+You can set NTH to send heartbeats to ASG in Queue Processor mode. This allows for a much longer grace period (up to 48 hours) for termination than the maximum heartbeat timeout of two hours.
+
```
I would add a line that explains when this feature would be useful: e.g. When a customer has pods that have long-running drain tasks.

Lu-David commented on README.md:
```diff
@@ -81,6 +81,78 @@ When using the ASG Lifecycle Hooks, ASG first sends the lifecycle action notific
+##### How to use
+
+- Configure a termination lifecycle hook on ASG (required). Set the heartbeat timeout value to be longer than the `Heartbeat Interval`. Each heartbeat signal resets this timeout, extending the duration that an instance remains in the `Terminating:Wait` state. Without this lifecycle hook, the instance will terminate immediately when termination event occurs.
+- Configure `Heartbeat Interval` (required) and `Heartbeat Until` (optional). NTH operates normally without heartbeats if neither value is set. If only the interval is specified, `Heartbeat Until` defaults to 172800 seconds (48 hours) and heartbeats will be sent. Providing both values enables NTH to run with heartbeats. `Heartbeat Until` must be provided with a valid `Heartbeat Interval`, otherwi ...
```
Plus 1 to this point

LikithaVemulapalli reviewed (approved):
/lgtm
